File: May-Challenge/week2/trie.js
/**
 * Initialize your data structure here.
 */
var Trie = function(v) {
    this.val = v;
    this.child = {};
    this.endHere = false;
    return this;
};

const root = new Trie('root');

/**
 * Inserts a word into the trie. 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let parent = root;
    for (const key in word.split('')) {
        if (parent.child[word[key]] === undefined) {
            parent.child[word[key]] = new Trie(word[key]);
        }
        parent = parent.child[word[key]];
        if (parseInt(key) === word.length-1) {
            parent.endHere = true;
        }
    }
};

/**
 * Returns if the word is in the trie. 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    parent = root;
    for (const char of word.split('')) {
        if (parent.child[char] === undefined) {
            return false;
        }
        parent = parent.child[char];
    }
    return parent.endHere;
};

/**
 * Returns if there is any word in the trie that starts with the given prefix. 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    parent = root;
    for (const char of prefix.split('')) {
        console.log(char)
        if (parent.child[char] === undefined) {
            return false;
        }
        parent = parent.child[char];
    }
    return true;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: May-Challenge/week2/trie.py
class TreeNode:

    def __init__(self, v):
        """
        Initialize your data structure here.
        """
        self.val = v
        self.child = {}
        self.endhere = False
        
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TreeNode(None)

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        parent = self.root
        for i, char in enumerate(word):
            if char not in parent.child:
                parent.child[char] = TreeNode(char)
            parent = parent.child[char]
            if i == len(word) - 1:
                parent.endhere = True
        

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        parent = self.root
        for char in word:
            if char not in parent.child:
                return False
            parent = parent.child[char]
        return parent.endhere
    
    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        parent = self.root
        for char in prefix:
            if char not in parent.child:
                return False
            parent = parent.child[char]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)